File: 3D_Flocking_pkg/boid.py
import numpy as np
import boidUtils as bu
import vectorUtils as vu

class Boid:

  def __init__(self, pos):
    self.pos = pos
    self.vel = np.random.random_sample(3) * 3
  
  def copyForOverflow(self):
    copy = Boid(self.pos.copy())
    copy.vel = self.vel.copy()
    return copy

  def update(self, maxSpeed, maxForce, secondsPast, boxSize):
    self.flockForceToApply = vu.limit(self.flockForceToApply, maxForce)
    # add force to current velocity
    self.vel += self.flockForceToApply * secondsPast
    self.vel = vu.limit(self.vel, maxSpeed)
    # and add the velocity to current position
    self.pos += self.vel * secondsPast
    # correct if the boid went past an edge
    self.pos[0] = bu.correctEdgeAxis(self.pos[0], 0, boxSize[0])
    self.pos[1] = bu.correctEdgeAxis(self.pos[1], 0, boxSize[1])
    self.pos[2] = bu.correctEdgeAxis(self.pos[2], 0, boxSize[2])

  def calcFlockForce(self, boids, percR, maxSpeed, maxForce, sizeBox, serpMult):
    bsInView = self.calcBoidsInView(boids, percR, sizeBox)
    s = self.calcSeparation(bsInView, percR, maxForce)
    c = self.calcCohesion(bsInView, maxSpeed, maxForce)
    a = self.calcAlignment(bsInView, maxSpeed, maxForce)
    # print(s)
    # print(c)
    # print(a)
    # apply modifiers
    s *= serpMult
    c *= .4
    a *= .2
    # add up and assign to flockForceToApply !!! flockForceToApply is reset, call applyDestinationSteer after this!
    self.flockForceToApply = s + c + a

  def calcBoidsInView(self, boids, percR, sizeBox):
    bs = []
    for b in boids:
      boid = bu.correctEdgeOverflowPerceptionR(self.pos, b, sizeBox, percR)
      if self is not boid and vu.dist(self.pos, boid.pos) <= percR:
        # print(boids[i])
        bs.append(boid)
        # print(bs[len(bs) - 1])
        # print("")
    
    return bs

  def calcSeparation(self, bsInView, percR, maxForce):
    if len(bsInView) == 0:
      return np.array((0, 0, 0), dtype=float)

    sep = np.array((0, 0, 0), dtype=float)
    for bInView in bsInView:
      pushForce = self.pos - bInView.pos
      l = vu.mag(pushForce)
      pushForce = vu.normalize(pushForce)
      pushForce *= 1 - (l / percR)
      sep += pushForce
    
    sep /= len(bsInView)
    sep *= maxForce
    return sep

  def calcAlignment(self, bsInView, maxSpeed, maxForce):
    if len(bsInView) == 0:
      return np.array((0, 0, 0), dtype=float)

    averageHeading = vu.normalize(self.vel)
    for bInView in bsInView:
      if vu.mag(bInView.vel) > 0:
        averageHeading += vu.normalize(bInView.vel)
    
    averageHeading /= len(bsInView) + 1
    return self.steerForceFromVector(averageHeading, maxSpeed, maxForce)

  def calcCohesion(self, bsInView, maxSpeed, maxForce):
    if len(bsInView) == 0:
      return np.array((0, 0, 0), dtype=float)

    centerOfMass = np.copy(self.pos)
    for bInView in bsInView:
      centerOfMass += bInView.pos
    
    centerOfMass /= len(bsInView) + 1
    return self.steerTowards(centerOfMass, maxSpeed, maxForce)

  def applyDestinationSteer(self, dest, maxSpeed, maxForce):
    seek = self.steerTowards(dest.copy(), maxSpeed, maxForce)
    seek *= .3
    self.flockForceToApply += seek

  def steerForceFromVector(self, desired, maxSpeed, maxForce):
    desired = vu.setMag(desired, maxSpeed)
    desired -= self.vel
    desired *= maxForce / maxSpeed
    return desired
  def steerTowards(self, dest, maxSpeed, maxForce):
    if np.array_equal(self.pos, dest):
      return np.array((0, 0, 0), dtype=float)
    dest -= self.pos
    return self.steerForceFromVector(dest, maxSpeed, maxForce)
  

File: Project1/Project1/Boid.cpp
#include "Vector3.h"
#include "BoidUtils.h"
#include "Boid.h"
#include <iostream>

using namespace std;

Boid::Boid(Vector3 p, Vector3 v)
{
	pos = p;
	vel = v;
}

void Boid::calcFlockForce(
	Boid* boids,
	int nBoids,
	float percR,
	float maxSpeed,
	float maxForce,
	Vector3 &sizeBox,
	float sepMultiplier) {

	Vector3 sep = Vector3(0, 0, 0);
	Vector3 ali = vel.normalized();
	Vector3 coh = pos.copy();

	float nBsInView = 0;
	for (int i = 0; i < nBoids; i++) {

		if (this != &boids[i]) {
			Boid bInView = BoidUtils::correctEdgeOverflowPerceptionR(pos, boids[i], sizeBox, percR);
			if ((pos - bInView.pos).length() <= percR) {
				nBsInView++;

				// separation
				Vector3 pushForce = pos - bInView.pos;
				float l = pushForce.length();
				pushForce.normalize();
				pushForce *= 1 - (l / percR);
				sep += pushForce;
				// alignment
				if (bInView.vel.length() > 0)
					ali += bInView.vel.normalized();

				// cohesion
				coh += bInView.pos;
			}
		}
	}
	if (nBsInView == 0) { flockForceToApply.makeTrivial(); return; }

	sep /= nBsInView;
	sep *= maxForce;

	ali /= nBsInView + 1;
	ali = steerForceFromVector(ali, maxSpeed, maxForce);

	coh /= nBsInView + 1;
	coh = steerTowards(coh, maxSpeed, maxForce);

	sep *= sepMultiplier;
	ali *= .2;
	coh *= .4;
	//sep.log();
	//ali.log();
	//coh.log();
	//std::cout << nBsInView << endl;
	flockForceToApply = sep + coh + ali;
}

void Boid::update(float maxSpeed, float maxForce, float timePast, Vector3 &boxSize) {

	flockForceToApply.limit(maxForce);
	vel += flockForceToApply * timePast;

	vel.limit(maxSpeed);
	pos += vel * timePast;

	pos.x = BoidUtils::correctEdgeAxis(pos.x, 0, boxSize.x);
	pos.y = BoidUtils::correctEdgeAxis(pos.y, 0, boxSize.y);
	pos.z = BoidUtils::correctEdgeAxis(pos.z, 0, boxSize.z);
}

Vector3& Boid::steerTowards(Vector3 &dest, float maxSpeed, float maxForce) {
	if (pos == dest) { dest = Vector3(0, 0, 0); return dest; }

	dest -= pos;
	return steerForceFromVector(dest, maxSpeed, maxForce);
}
Vector3& Boid::steerForceFromVector(Vector3 &desired, float maxSpeed, float maxForce) {
	desired.setMag(maxSpeed);
	desired -= vel;
	desired *= (maxForce / maxSpeed);
	return desired;
}
